lib/ContextChat/ContextChatProvider.php

```diff
@@ -0,0 +1,141 @@
+
+/**
+ * @implements IEventListener<Event>
+ */
+class ContextChatProvider implements IContentProvider, IEventListener {
+
+	public function __construct(
+		private JobsService $jobsService,
+		private AccountService $accountService,
+		private MailManager $mailManager,
+		private MessageMapper $messageMapper,
+		private IURLGenerator $urlGenerator,
+		private IUserManager $userManager,
+		private IContentManager $contentManager,
+		private IJobList $jobList,
+	) {
+	}
+
+	public function handle(Event $event): void {
+		if (!$this->contentManager->isContextChatAvailable()) {
+			return;
+		}
+
+		if ($event instanceof ContentProviderRegisterEvent) {
+			$event->registerContentProvider($this->getAppId(), $this->getId(), self::class);
+			return;
+		}
+
+		if ($event instanceof NewMessagesSynchronized) {
+			$messageIds = array_map(static fn (Message $m): int => $m->getId(), $event->getMessages());
+
+			// Ensure that there are messages to sync
+			if (count($messageIds) === 0) {
+				return;
+			}
+
+			$userId = $event->getAccount()->getUserId();
+			$accountId = $event->getAccount()->getId();
+			$mailboxId = $event->getMailbox()->getId();
+
+			$this->jobsService->updateOrCreate($userId, $accountId, $mailboxId, min($messageIds));
+			return;
+		}
+
+		if ($event instanceof MessageDeletedEvent) {
+			$this->contentManager->deleteContent($this->getAppId(), $this->getId(), [strval($event->getMessageId())]);
+			return;
+		}
+	}
+
+	/**
+	 * The ID of the provider
+	 *
+	 * @return string
+	 * @since 5.2.0
+	 */
+	public function getId(): string {
+		return 'mail';
+	}
+
+	/**
+	 * The ID of the app making the provider avaialble
+	 *
+	 * @return string
+	 * @since 5.2.0
+	 */
+	public function getAppId(): string {
+		return Application::APP_ID;
+	}
+
+	/**
+	 * The absolute URL to the content item
+	 *
+	 * @param string $id
+	 * @return string
+	 * @since 5.2.0
+	 */
+	public function getItemUrl(string $id): string {
+		// Get mailbox ID from message ID
+		$messages = $this->messageMapper->findByIds('', [(int)$id], '');
+		if (!$messages) {
+			return '';
+		}
+		$mailboxId = $messages[0]->getMailboxId();
+
+		return $this->urlGenerator->linkToRouteAbsolute('mail.page.thread', [ 'mailboxId' => $mailboxId, 'id' => $id ]);
+	}
+
+	/**
+	 * Starts the initial import of content items into context chat
+	 *
+	 * @return void
+	 * @since 5.2.0
+	 */
+	public function triggerInitialImport(): void {
+		$this->userManager->callForSeenUsers(function (IUser $user): void {
+			$userId = $user->getUID();
+			$userAccounts = $this->accountService->findByUserId($userId);
+
+			foreach ($userAccounts as $account) {
+				$mailboxes = $this->mailManager->getMailboxes($account);
+
+				foreach ($mailboxes as $mailbox) {
+					$messageIds = $this->messageMapper->findAllIds($mailbox);
+
+					if (count($messageIds) > 0) {
+						$this->jobsService->updateOrCreate($userId, $account->getId(), $mailbox->getId(), 0);
+					}
+				}
+			}
+		});
+	}
```

Comment:
Has this been tested with moderate to large setups? I'm afraid we will run out of memory

You are loading all IDs of a mailbox without really using the result. Again, this has to be pushed to the db with a limit=1.

Reply:
Ideally, we'd check if there is at least one message that satisfies all conditions for import (e.g. less than a year old), but for simplicity, we can check if there is at least one message in the mailbox, period. For that, I'm guessing we can use `MessageMapper::findHighestUid` and check for `null` rather than use `findAllIds`.